Fix #1350: redesign the profile page

kerryxwang opened the pull request:
New pull request addressing comments from pull request at new-profile-page branch.

kerryxwang commented:
@seanlip @wxyxinyu This is the new pull request after I got the git push to go through!

kerryxwang commented:
The last task to be done after this is to implement in-place editing.

wxyxinyu commented:
Hi @kerryxwang, I checked out the branch, but I found some issues:

I think there was an issue with the merge: the changes to the navbar seem to be broken.
![oppia gallery](https://cloud.githubusercontent.com/assets/5061134/14092915/1d2bedf2-f518-11e5-891e-ab63b2ccbda8.png)

I also got this console error when looking at the profile page, while the author hasn't created anything yet.
![screenshot from 2016-03-28 19 04 52](https://cloud.githubusercontent.com/assets/5061134/14092907/1055f03c-f518-11e5-8024-2a312bd24fe7.png)
It works fine, and looks nice, when the author creates at least one exploration.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -1026,8 +1089,8 @@ md-card.oppia-gallery-tile {
 }
 
 .oppia-html-select .oppia-html-select-option oppia-noninteractive-image {
```
There's a whole bunch of stuff here that seems to have been deleted/changed, but shouldn't be. Please could you reinstate it?

kerryxwang replied:
I've reinstated it, but are you sure they're not supposed to be deleted? All these changes were taken when I merged develop with upstream...

kerryxwang replied:
I do not feel comfortable reinstating the changes. As I did not make the changes, I am not sure which changes to keep and which to changes to revert.

seanlip replied:
OK, here is another way to look at it.

The view [here](https://github.com/oppia/oppia/pull/1638/files) shows the changes your PR makes to oppia:develop. If you don't mean to make changes to classes like .oppia-html-select or similar, then there should be no change shown there. (But there is, so there's a problem.)

That is to say, when you've done it correctly, that diff view should only reflect the changes you've deliberately made, and nothing else.

seanlip replied:
(Does that help?)

kerryxwang replied:
Yes.

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/profile/Profile.js:
```diff
@@ -29,22 +31,96 @@ oppia.controller('Profile', [
+      });
+
+      $scope.currentPageNumber = 0;
+      $scope.pageSize = 6;
```
This is a constant, right? If so, $scope.PAGE_SIZE

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/profile/Profile.js:
```diff
@@ -29,22 +31,96 @@ oppia.controller('Profile', [
+      $scope.pageSize = 6;
+      $scope.startingExplorationNumber = 1;
+      $scope.endingExplorationNumber = 6;
+      $scope.updateDisplayedExplorationNumbers = function() {
```
In the directive template, could you just do the following:

    <[currentPageNumber * pageSize + 1]>

and

    <[Math.min(userEditedExplorations.length, (currentPageNumber + 1) * pageSize)]>

Then you don't need this function.

kerryxwang replied:
I was able to take out the function completely.

seanlip commented on core/templates/dev/head/components/ExplorationSummaryTileMobileDirective.js:
```diff
@@ -0,0 +1,140 @@
+      isCommunityOwned: '&isCommunityOwned'
+    },
+    templateUrl: 'summaryTile/exploration-mobile',
+    link: function(scope, element) {
```
Way too much duplication here (with the non-mobile directive). If we keep this, people will change the code for one of them, and forget to change the other.

Instead, can you add an optional attribute to the existing explorationSummaryTile directive -- something like mobile-cutoff? Then change its HTML template to show the small version if the screen width is below mobile-cutoff, and the large version otherwise. If mobile-cutoff is not specified (as is the case on existing uses of explorationSummaryTile), always show the large version.

kerryxwang replied:
How do we add an optional attribute? Is this similar to the process I used to make the profile page responsive?

seanlip commented on core/templates/dev/head/profile/Profile.js:
```diff
@@ -14,6 +14,8 @@
 /**
  * @fileoverview Data and controllers for the Oppia profile page.
+ *
+ * @author sean@seanlip.org (Sean Lip)
```
Please remove these two lines -- we've removed author lines throughout the codebase, since files tend to be edited by multiple people.

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/profile/profile.html:
```diff
@@ -31,104 +31,91 @@
+          Previous
+        </button>
+        Showing <[startingExplorationNumber]> - <[endingExplorationNumber]> of <[numUserPortfolioExplorations]>
+        <button ng-disabled="(currentPageNumber + 1) * PAGE_SIZE >= userEditedExplorations.length" ng-click="goToNextPage();">
```
Remove semicolon at end (in ng-click) -- it's not needed.

kerryxwang replied:
Done.

seanlip commented:
A couple of UI issues:

In the gallery, the tiles look wrong. Not sure what happened there, but the display of the summary tiles in other pages should not be affected.

![gallery-tiles-wrong](https://cloud.githubusercontent.com/assets/10575562/14265744/26b048a4-fa78-11e5-9caa-98a2d34c211e.png)

Also, when you zoom into the profile page, and hover over one of the tiles, a vertical scrollbar appears in the middle of the page, which is undesirable:

![scrollbar-mobile](https://cloud.githubusercontent.com/assets/10575562/14265776/4dae91fe-fa78-11e5-996c-7f197cfe1592.png)

This is all I've managed to detect so far. Could you take a look?

seanlip commented on core/templates/dev/head/components/ExplorationSummaryTileDirective.js:
```diff
@@ -134,6 +137,16 @@ oppia.directive('explorationSummaryTile', [function() {
           return result;
         };
+
+        $scope.isWindowLarge = (
```
Need to handle the case when mobileCutoffPx is not specified (if so, set it to 0 explicitly).

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/components/ExplorationSummaryTileDirective.js:
```diff
@@ -34,7 +34,8 @@ oppia.directive('explorationSummaryTile', [function() {
       openInNewWindow: '@openInNewWindow',
-      isCommunityOwned: '&isCommunityOwned'
+      isCommunityOwned: '&isCommunityOwned',
+      mobileCutoffPx: '@mobileCutoffPx'
```
Add comment: 

// If the screen width is below the threshold defined here, the mobile version of the summary tile is displayed. This attribute is optional: if it is not specified, it is treated as 0, which means that the desktop version of the summary tile is always displayed.

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/components/exploration_summary_tile_directive.html:
```diff
@@ -1,9 +1,14 @@
+        <h2 ng-if="isWindowLarge" class="exploration-title protractor-test-exp-summary-tile-title">
+          <[getExplorationTitle()]>
+        </h2>
+        <h2 ng-if="isWindowLarge==false" class="exploration-title protractor-test-exp-summary-tile-title">
```
Use !isWindowLarge instead.

seanlip replied:
Actually, you can do:
```
      <h2 class="exploration-title protractor-test-exp-summary-tile-title">
        <span ng-if="isWindowLarge"><[getExplorationTitle()]></span>
        <span ng-if="!isWindowLarge"><[getExplorationTitle()|truncate:32]></span>
      </h2>
```

kerryxwang replied:
Done.

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/components/exploration_summary_tile_directive.html:
```diff
@@ -35,9 +40,9 @@ <h2 class="exploration-title protractor-test-exp-summary-tile-title"><[getExplor
         </div>
       </div>
-      <div class="top-section-mask mask"></div>
+      <div class="title-section-mask mask"></div>
```
Where is .mask defined?

kerryxwang replied:
It is not defined, I deleted the second "mask"

seanlip commented on core/templates/dev/head/components/exploration_summary_tile_directive.html:
```diff
@@ -72,6 +77,28 @@ <h2 class="exploration-title protractor-test-exp-summary-tile-title"><[getExplor
           </li>
         </ul>
       </div>
+      <div ng-if="isWindowLarge==false" class="right-section">
```
!isWindowLarge

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/components/exploration_summary_tile_directive.html:
```diff
@@ -72,6 +77,28 @@ <h2 class="exploration-title protractor-test-exp-summary-tile-title"><[getExplor
       </div>
+      <div ng-if="isWindowLarge==false" class="right-section">
+        <ul class="metrics">
+          <li>
```
There's a lot of duplication here. Can this be reduced? If the challenge is in conditionally setting the values of  `layout` and `layout-align`, consider ng-attr:

    http://stackoverflow.com/questions/15696416/what-is-the-best-way-to-conditionally-apply-attributes-in-angular

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/components/exploration_summary_tile_directive.html:
```diff
@@ -235,6 +262,15 @@ <h2 class="exploration-title protractor-test-exp-summary-tile-title"><[getExplor
     opacity: 0.5;
   }
 
+   .oppia-exploration-summary-tile .right-section {
```
indentation of this line is wrong

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -110,144 +110,195 @@ body {
-    margin-right: 10px;
-  }
+.oppia-profile-container {
+  margin-left: auto;
```
A simplification is:

    margin: 60px auto 30px auto;

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -110,144 +110,195 @@ body {
-.oppia-profile-container, .oppia-dashboard-container {
+.oppia-dashboard-container {
   max-width: 800px;
   margin-left: auto;
```
A simplification is:

    margin: 60px auto 30px auto;

seanlip replied:
(i.e. top, right, bottom, left)

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -1128,6 +1180,13 @@ md-card.oppia-gallery-tile {
   }
 }
 
+.oppia-alert-blocks {
```
What is this referring to?

kerryxwang replied:
I do not know, this was what I reinstated from my previous push that somehow deleted this code.

seanlip replied:
OK -- if you didn't mean to add it, I would suggest deleting it. The file diff at https://github.com/oppia/oppia/pull/1638/ should only reflect the changes you've added.

seanlip commented on core/templates/dev/head/profile/Profile.js:
```diff
@@ -29,22 +31,87 @@ oppia.controller('Profile', [
+      $scope.username = {
+        title: 'Username',
+        value: data.username,
+        helpText: (data.username)
```
Why do we need the help text in the first place? It seems a bit odd in the UI -- it gives no additional information.

seanlip replied:
(If it's an issue with non-usernames, maybe we should only show it if the username is too long?)

kerryxwang replied:
How do we only show the helpText in the scenario that the username is too long?

seanlip replied:
Use an ng-if on the popover, so that it's only displayed in the right circumstances. You'll probably want to compute a $scope.usernameIsLong variable here for that purpose.

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/profile/Profile.js:
```diff
@@ -29,22 +31,87 @@ oppia.controller('Profile', [
+
+      $scope.currentPageNumber = 0;
+      $scope.PAGE_SIZE = 6;
+      $scope.startingExplorationNumber = 1;
```
This and the following line appear to get set, but are never changed. This seems like an error...

kerryxwang replied:
I change this number using directives.

seanlip commented on core/templates/dev/head/profile/Profile.js:
```diff
@@ -29,22 +31,87 @@ oppia.controller('Profile', [
+          console.log('Error: cannot decrement page');
+        } else {
+          $scope.currentPageNumber--;
+          $scope.updateDisplayedExplorationNumbers();
```
Where is this defined?

kerryxwang replied:
I took out this function and used directives to change the displayed exploration numbers.

seanlip commented on core/templates/dev/head/profile/profile.html:
```diff
@@ -31,104 +31,91 @@
+            ratings="expl.ratings"
+            thumbnail-icon-url="expl.thumbnail_icon_url"
+            thumbnail-bg-color="expl.thumbnail_bg_color"
+            num-views="expl.num_views"
```
add is-community-editable="expl.community_editable"

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/profile/profile.html:
```diff
@@ -31,104 +31,91 @@
+          </exploration-summary-tile>
+
+        <div ng-if="userEditedExplorations.length === 0">
+          <em></br>This user hasn't created or edited any explorations yet.</em>
```
Why the /br? If you do want a break, it should probably go outside the em, and be formatted as just `<br>`.

kerryxwang replied:
The break is for design.

seanlip replied:
Can it be replaced with a margin-top on the div containing it? In general, for spacing-related stuff, it's better to use CSS, especially since this isn't really semantically a line break.

seanlip commented:
Hi @kerryxwang -- I've taken a review pass. Unless i'm missing something, it looks to me as if there are serious errors in the paging logic. Did you test manually?

seanlip commented on core/templates/dev/head/profile/profile.html:
```diff
@@ -31,104 +31,91 @@
+    <md-card class="oppia-profile-content-card">
+      <strong>Portfolio</strong>
+      <md-content class="oppia-profile-portfolio-container">
+          <exploration-summary-tile ng-repeat="expl in getExplorationsToDisplay()"
```
Indentation off -- deindent this block by 2.

seanlip replied:
@kerryxwang -- I think you missed this?

kerryxwang replied:
Done.

seanlip commented:
Hi @kerryxwang -- just following up, how is this going? Thanks!

seanlip commented:
@kerryxwang -- ping! (No worries if you're busy, but please let me know either way -- we're hoping to ship the profile page in the end-of-April release.)

Thanks!

kerryxwang commented:
Hi @seanlip, I apologize the most recent PR took so long!

seanlip commented:
No worries! Just to double check, do you think it is ready for review and merge now?

kerryxwang commented:
I believe so!

seanlip commented on core/templates/dev/head/components/ExplorationSummaryTileDirective.js:
```diff
@@ -134,6 +141,21 @@ oppia.directive('explorationSummaryTile', [function() {
+        if ($scope.mobileCutoffPx == null) {
+          $scope.mobileCutoffPx = 0;
+        }
+        console.log($scope.mobileCutoffPx);
```
Delete this

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/components/ExplorationSummaryTileDirective.js:
```diff
@@ -134,6 +141,21 @@ oppia.directive('explorationSummaryTile', [function() {
+        console.log($scope.mobileCutoffPx);
+        $scope.isWindowLarge = (
+          windowDimensionsService.getWidth() >= $scope.mobileCutoffPx);
+        console.log($scope.isWindowLarge);
```
Ditto

kerryxwang replied:
Done.

seanlip commented on core/templates/dev/head/profile/profile.html:
```diff
@@ -31,104 +31,93 @@
       </md-content>
+
+      <span class="oppia-profile-portfolio-pages" ng-if="numUserPortfolioExplorations > 6">
+        <button ng-disabled="currentPageNumber == 0" ng-click="goToPreviousPage()">
```
These buttons don't really match the style of the rest of the site. Instead of using buttons, could you use the material design left/right arrows, similar to what we do for the progress dots in the learner view?

kerryxwang replied:
By learned view, do you mean the buttons in the explorations?
https://www.oppia.org/explore/9

kerryxwang replied:
Done.

seanlip commented:
Hi @kerryxwang -- this is looking great. I just have a few more comments (please note [this earlier one](https://github.com/oppia/oppia/pull/1638#discussion_r60879920) as well).

In addition, two general things:
- The tiles in the gallery page look totally messed up on narrow screen widths. I think something is not working correctly with the mobile cutoff -- could you please debug?
- Create a profile with 6 explorations, and make the screen width approximately 900px; you'll end up with something like the following screenshot. Could you please fix the footer position? You may want to look at how /about, /participate are structured -- they are long pages, but they don't run into this issue.

    ![bad-footer](https://cloud.githubusercontent.com/assets/10575562/14778694/3e4a858a-0a89-11e6-94a4-5ebf693a19c6.png)

Once these are sorted out, this PR can be merged. Thanks!

kerryxwang commented:
Hi Sean, I addressed the PR comments and the 2 major bugs. I also changed the design of the mobile tiles a little bit by including the recently updated icon/update date. Please take a look and let me know if it looks too cramped to you. If not, I think we should keep it!

seanlip commented:
Hi @kerryxwang -- this looks amazing. LGTM -- and thank you! I'll merge when the Travis checks pass, and it'll go out in the upcoming release. :-)

seanlip commented:
Update: Travis-CI checks fail, but this is due to an extra ">" character in activity_tiles_infinity_grid.html, and with that fixed, they pass (see #1768). Based on this confirmation, I'm going to merge this into develop and fix the ">" character post-hoc so that it's possible to squash the commits.

kerryxwang commented:
Thank you Sean! :)

On Thu, Apr 28, 2016 at 8:28 PM, Sean Lip <notifications@github.com> wrote:

> Hi @kerryxwang <https://github.com/kerryxwang> -- this looks amazing.
> LGTM -- and thank you! I'll merge when the Travis checks pass, and it'll go
> out in the upcoming release. :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/oppia/oppia/pull/1638#issuecomment-215618551>
>
